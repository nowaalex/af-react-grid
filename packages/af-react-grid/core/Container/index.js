import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import { cx as _cx } from "emotion";
import { css as _css } from "emotion";

function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }

import React, { forwardRef } from "react";
import { TypeContext } from "../contexts";

var containerBaseCss = /*#__PURE__*/_css(process.env.NODE_ENV === "production" ? {
  name: "1dsbrkv-containerBaseCss",
  styles: "display:flex;flex-wrap:nowrap;;label:containerBaseCss;"
} : {
  name: "1dsbrkv-containerBaseCss",
  styles: "display:flex;flex-wrap:nowrap;;label:containerBaseCss;",
  map: "/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uL3NyYy9Db250YWluZXIvaW5kZXguanMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBSzRCIiwiZmlsZSI6Ii4uLy4uL3NyYy9Db250YWluZXIvaW5kZXguanMiLCJzb3VyY2VzQ29udGVudCI6WyJcbmltcG9ydCBSZWFjdCwgeyBmb3J3YXJkUmVmIH0gZnJvbSBcInJlYWN0XCI7XG5pbXBvcnQgeyBUeXBlQ29udGV4dCB9IGZyb20gXCIuLi9jb250ZXh0c1wiO1xuaW1wb3J0IHsgY3NzLCBjeCB9IGZyb20gXCJlbW90aW9uXCI7XG5cbmNvbnN0IGNvbnRhaW5lckJhc2VDc3MgPSBjc3NgXG4gICAgZGlzcGxheTogZmxleDtcbiAgICBmbGV4LXdyYXA6IG5vd3JhcDtcbmA7XG5cbmNvbnN0IGNvbENzcyA9IGN4KGNvbnRhaW5lckJhc2VDc3MsIGNzc2BmbGV4LWRpcmVjdGlvbjogY29sdW1uO2ApO1xuXG5jb25zdCByb3dDc3MgPSBjeChjb250YWluZXJCYXNlQ3NzLCBjc3NgZmxleC1kaXJlY3Rpb246IHJvdztgKTtcblxuXG5jb25zdCBDb250YWluZXIgPSBmb3J3YXJkUmVmKCh7IHR5cGUsIENvbXBvbmVudCA9IFwiZGl2XCIsIGNsYXNzTmFtZSwgLi4ucHJvcHMgfSwgcmVmICkgPT4gKFxuICAgIDxUeXBlQ29udGV4dC5Qcm92aWRlciB2YWx1ZT17dHlwZX0+XG4gICAgICAgIDxDb21wb25lbnQgY2xhc3NOYW1lPXtjeCh0eXBlPT09XCJjb2xcIj9jb2xDc3M6cm93Q3NzLGNsYXNzTmFtZSl9IHsuLi5wcm9wc30gcmVmPXtyZWZ9IC8+XG4gICAgPC9UeXBlQ29udGV4dC5Qcm92aWRlcj5cbikpO1xuXG5leHBvcnQgZGVmYXVsdCBDb250YWluZXI7Il19 */",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
});

var colCss = _cx(containerBaseCss, /*#__PURE__*/_css(process.env.NODE_ENV === "production" ? {
  name: "7rtuif-colCss",
  styles: "flex-direction:column;;label:colCss;"
} : {
  name: "7rtuif-colCss",
  styles: "flex-direction:column;;label:colCss;",
  map: "/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uL3NyYy9Db250YWluZXIvaW5kZXguanMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBVXVDIiwiZmlsZSI6Ii4uLy4uL3NyYy9Db250YWluZXIvaW5kZXguanMiLCJzb3VyY2VzQ29udGVudCI6WyJcbmltcG9ydCBSZWFjdCwgeyBmb3J3YXJkUmVmIH0gZnJvbSBcInJlYWN0XCI7XG5pbXBvcnQgeyBUeXBlQ29udGV4dCB9IGZyb20gXCIuLi9jb250ZXh0c1wiO1xuaW1wb3J0IHsgY3NzLCBjeCB9IGZyb20gXCJlbW90aW9uXCI7XG5cbmNvbnN0IGNvbnRhaW5lckJhc2VDc3MgPSBjc3NgXG4gICAgZGlzcGxheTogZmxleDtcbiAgICBmbGV4LXdyYXA6IG5vd3JhcDtcbmA7XG5cbmNvbnN0IGNvbENzcyA9IGN4KGNvbnRhaW5lckJhc2VDc3MsIGNzc2BmbGV4LWRpcmVjdGlvbjogY29sdW1uO2ApO1xuXG5jb25zdCByb3dDc3MgPSBjeChjb250YWluZXJCYXNlQ3NzLCBjc3NgZmxleC1kaXJlY3Rpb246IHJvdztgKTtcblxuXG5jb25zdCBDb250YWluZXIgPSBmb3J3YXJkUmVmKCh7IHR5cGUsIENvbXBvbmVudCA9IFwiZGl2XCIsIGNsYXNzTmFtZSwgLi4ucHJvcHMgfSwgcmVmICkgPT4gKFxuICAgIDxUeXBlQ29udGV4dC5Qcm92aWRlciB2YWx1ZT17dHlwZX0+XG4gICAgICAgIDxDb21wb25lbnQgY2xhc3NOYW1lPXtjeCh0eXBlPT09XCJjb2xcIj9jb2xDc3M6cm93Q3NzLGNsYXNzTmFtZSl9IHsuLi5wcm9wc30gcmVmPXtyZWZ9IC8+XG4gICAgPC9UeXBlQ29udGV4dC5Qcm92aWRlcj5cbikpO1xuXG5leHBvcnQgZGVmYXVsdCBDb250YWluZXI7Il19 */",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
}));

var rowCss = _cx(containerBaseCss, /*#__PURE__*/_css(process.env.NODE_ENV === "production" ? {
  name: "10yfcpt-rowCss",
  styles: "flex-direction:row;;label:rowCss;"
} : {
  name: "10yfcpt-rowCss",
  styles: "flex-direction:row;;label:rowCss;",
  map: "/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uL3NyYy9Db250YWluZXIvaW5kZXguanMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBWXVDIiwiZmlsZSI6Ii4uLy4uL3NyYy9Db250YWluZXIvaW5kZXguanMiLCJzb3VyY2VzQ29udGVudCI6WyJcbmltcG9ydCBSZWFjdCwgeyBmb3J3YXJkUmVmIH0gZnJvbSBcInJlYWN0XCI7XG5pbXBvcnQgeyBUeXBlQ29udGV4dCB9IGZyb20gXCIuLi9jb250ZXh0c1wiO1xuaW1wb3J0IHsgY3NzLCBjeCB9IGZyb20gXCJlbW90aW9uXCI7XG5cbmNvbnN0IGNvbnRhaW5lckJhc2VDc3MgPSBjc3NgXG4gICAgZGlzcGxheTogZmxleDtcbiAgICBmbGV4LXdyYXA6IG5vd3JhcDtcbmA7XG5cbmNvbnN0IGNvbENzcyA9IGN4KGNvbnRhaW5lckJhc2VDc3MsIGNzc2BmbGV4LWRpcmVjdGlvbjogY29sdW1uO2ApO1xuXG5jb25zdCByb3dDc3MgPSBjeChjb250YWluZXJCYXNlQ3NzLCBjc3NgZmxleC1kaXJlY3Rpb246IHJvdztgKTtcblxuXG5jb25zdCBDb250YWluZXIgPSBmb3J3YXJkUmVmKCh7IHR5cGUsIENvbXBvbmVudCA9IFwiZGl2XCIsIGNsYXNzTmFtZSwgLi4ucHJvcHMgfSwgcmVmICkgPT4gKFxuICAgIDxUeXBlQ29udGV4dC5Qcm92aWRlciB2YWx1ZT17dHlwZX0+XG4gICAgICAgIDxDb21wb25lbnQgY2xhc3NOYW1lPXtjeCh0eXBlPT09XCJjb2xcIj9jb2xDc3M6cm93Q3NzLGNsYXNzTmFtZSl9IHsuLi5wcm9wc30gcmVmPXtyZWZ9IC8+XG4gICAgPC9UeXBlQ29udGV4dC5Qcm92aWRlcj5cbikpO1xuXG5leHBvcnQgZGVmYXVsdCBDb250YWluZXI7Il19 */",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
}));

var Container = forwardRef(function (_ref, ref) {
  var type = _ref.type,
      _ref$Component = _ref.Component,
      Component = _ref$Component === void 0 ? "div" : _ref$Component,
      className = _ref.className,
      props = _objectWithoutPropertiesLoose(_ref, ["type", "Component", "className"]);

  return /*#__PURE__*/React.createElement(TypeContext.Provider, {
    value: type
  }, /*#__PURE__*/React.createElement(Component, _extends({
    className: _cx(type === "col" ? colCss : rowCss, className)
  }, props, {
    ref: ref
  })));
});
export default Container;