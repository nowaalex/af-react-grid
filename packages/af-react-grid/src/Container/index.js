
import React, { forwardRef } from "react";
import { TypeContext } from "../contexts";
import { css, cx } from "emotion";

const containerBaseCss = css`
    display: flex;
    flex-wrap: nowrap;
`;

const colCss = cx(containerBaseCss, css`flex-direction: column;`);

const rowCss = cx(containerBaseCss, css`flex-direction: row;`);


const Container = forwardRef(({ type, Component = "div", className, ...props }, ref ) => (
    <TypeContext.Provider value={type}>
        <Component className={cx(type==="col"?colCss:rowCss,className)} {...props} ref={ref} />
    </TypeContext.Provider>
));

export default Container;