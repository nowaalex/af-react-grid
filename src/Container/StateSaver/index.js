import { UNIQUE_HASH } from "../../constants";

class StateSaver{

    constructor(){

        const obj = localStorage.getItem( UNIQUE_HASH );
        this.StorageObject = obj ? JSON.parse( obj ) : {};

        window.addEventListener( "beforeunload", () => {
            if( Object.keys( this.StorageObject ).length > 1 ){
                localStorage.setItem(
                    UNIQUE_HASH,
                    JSON.stringify( this.StorageObject )
                );
            }
        });
    }

    addStylesInfo( keyName, stylesObj ){
        this.StorageObject[ keyName ] = stylesObj;
    }

    getStylesInfo( keyName ){
        return this.StorageObject[ keyName ] || {};
    }
}

export default new StateSaver;