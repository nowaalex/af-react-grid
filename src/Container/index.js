import React from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";
import Resizer from "../Resizer";

import _ from "lodash";

const ByType = {
    row: {
        style: {
            display: "flex",
            flexFlow: "row nowrap"
        },
        ptr: "pageX",
        dim: "clientWidth",
        prop: "width",
        min: "minWidth",
        max: "maxWidth"
    },
    col: {
        style: {
            display: "flex",
            flexFlow: "column nowrap"
        },
        ptr: "pageY",
        dim: "clientHeight",
        prop: "height",
        min: "minHeight",
        max: "maxHeight"
    }
}

class Container extends React.Component{

    static propTypes = {
        type:                   PropTypes.oneOf([ "row", "col" ]).isRequired,
        resizerClassName:       PropTypes.string,
        children:               PropTypes.node,
    }

    state = {}

    /* Initial dimensions caches, used to calculate sizes onDrag */
    _initPtrPageDist;
    _curD1;
    _curD2;
    _minD1;
    _maxD1;
    _minD2;
    _maxD2;

    refsArr = [];

    _setInitialDimensionsCache( elIndex, fieldIndex ){

        const el = this.refsArr[ elIndex ];

        const { max, min, dim } = ByType[ this.props.type ];
        
        const obj = getComputedStyle( el );

        this[ "_curD" + fieldIndex ] = el[ dim ];

        const minVal = parseInt( obj[ min ], 10 );
        this[ "_minD" + fieldIndex ] = isNaN( minVal ) ? 0 : minVal;

        const maxVal = parseInt( obj[ max ], 10 );
        this[ "_maxD" + fieldIndex ] = isNaN( maxVal ) ? null : maxVal;
    }

    onStart = ( index, e ) => {

        if( process.env.NODE_ENV === "development" ){
            console.log( "onStart", index, e );
        }

        const { ptr } = ByType[ this.props.type ];

        this._initPtrPageDist = e[ ptr ];

        this._setInitialDimensionsCache( index - 1, 1 );
        this._setInitialDimensionsCache( index + 1, 2 );

        const sum = this._curD1 + this._curD2;

        if( this._maxD1 === null ){
            this._maxD1 = sum - this._minD2; 
        }

        if( this._maxD2 === null ){
            this._maxD2 = sum - this._minD1;
        }
    }

    _getChangedState( curState, prop, index, step ){
        return ({
            [ index - 1 ]: {
                ...curState[ index - 1 ],
                [prop]: _.clamp(
                    this._curD1 + step,
                    this._minD1,
                    this._maxD1
                )
            },
            [ index + 1 ]: {
                ...curState[ index + 1 ],
                [prop]: _.clamp(
                    this._curD2 - step,
                    this._minD2,
                    this._maxD2
                )
            }
        })
    }

    onDrag = ( index, e ) => {

        if( process.env.NODE_ENV === "development" ){
            console.log( "onDrag", index, e );
        }

        const { ptr, prop } = ByType[ this.props.type ];
        const step = e[ ptr ] - this._initPtrPageDist;

        this.setState( curState => this._getChangedState(
            curState,
            prop,
            index,
            step
        ));
    }

    /* by default only first arg is used as memoization key */
    _getSaveRef = _.memoize( index => node => {
        this.refsArr[ index ] = node ? ReactDOM.findDOMNode( node ) : null;
    })

    childrenMapper = ( el, index ) => {

        if( !el ){
            return el;
        }

        const { type, props } = el;

        if( type === React.Fragment ){
            throw new Error( "Fragments are not supported in ResizableFlexGrid right now." );
        }

        return React.cloneElement( el, type === Resizer ? {
            index,
            onDrag: this.onDrag,
            onStart: this.onStart,
            type: this.props.type,
            ref: this._getSaveRef( index )
        } : {
            style: props.style ? {
                ...props.style,
                ...this.state[ index ]
            } : this.state[ index ],
            ref: this._getSaveRef( index )
        })
    }

    render(){

        const {
            type,
            className,
            children,
            style
        } = this.props;

        const baseStyle = ByType[ type ].style;

        return (
            <div
                style={ style ? { ...baseStyle, ...style } : baseStyle }
                className={className}
                children={React.Children.map( children, this.childrenMapper )}
            />
        )
    }

    _dimensionsStateModifier = ( curState, { type } ) => {
        
        const { prop, dim } = ByType[ type ];

        return this.refsArr.reduce(( acc, ref, i ) => {
            if( ref ){
                acc[ i ] = {
                    ...curState[ i ],
                    [prop]: ref[ dim ]
                }
            }
            return acc;
        }, {});
    }

    setExactDimensions = _.throttle(() => this.setState( this._dimensionsStateModifier ), 150, {
        leading: false
    });

    componentDidMount(){
        this.setExactDimensions();
        window.addEventListener( "resize", this.setExactDimensions );
    }

    componentDidUpdate({
        children: prevChildren,
        height: prevHeight,
        width: prevWidth
    }){

        const {
            children, 
            height,
            width
        } = this.props;

        const prevChildrenLen = React.Children.count( prevChildren );
        const curChildrenLen =  React.Children.count( children );

        if(
            width !== prevWidth ||
            height !== prevHeight ||
            prevChildrenLen !== curChildrenLen
        ){
            if( prevChildrenLen > curChildrenLen ){
                this.refsArr.splice( curChildrenLen );
            }
            this.setExactDimensions();
        }
    }

    componentWillUnmount(){
        this.setExactDimensions.cancel();
        window.removeEventListener( "resize", this.setExactDimensions );
    }
}

export default Container;