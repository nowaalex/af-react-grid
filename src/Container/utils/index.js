import {
    flexColumnNowrap,
    flexRowNowrap
} from "flexclasses.module";

export const ByType = {
    row: {
        className: flexRowNowrap,
        ptr: "pageX",
        dim: "clientWidth",
        prop: "width",
        min: "minWidth",
        max: "maxWidth"
    },
    col: {
        className: flexColumnNowrap,
        ptr: "pageY",
        dim: "clientHeight",
        prop: "height",
        min: "minHeight",
        max: "maxHeight"
    }
}