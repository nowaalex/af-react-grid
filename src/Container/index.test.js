import React from "react";
import Container from "../Container";
import renderer from "react-test-renderer";
import mockConsole from "jest-mock-console";

test( "Empty container renders correctly", () => {

    const component = renderer.create(
        <Container />
    );

    let tree = component.toJSON();
    expect( tree ).toMatchSnapshot();
});

test( "Allows rendering of bools, undefined, null and empty strings", () => {

    const component = renderer.create(
        <Container>
            {false}
            {undefined}
            {null}
            
        </Container>
    );

    let tree = component.toJSON();
    expect( tree ).toMatchSnapshot();
});

test( "Does not allow Fragments inside Container", () => {

    const restoreConsole = mockConsole();

    renderer.create(
        <Container>
            <React.Fragment />
        </Container>
    );
    
    expect( console.error ).toHaveBeenCalled();

    restoreConsole();
});

test( "Does not allow arrays inside Container", () => {

    const restoreConsole = mockConsole();

    renderer.create(
        <Container>
            {[
                <div>1</div>
            ]}
        </Container>
    );
    
    expect( console.error ).toHaveBeenCalled();

    restoreConsole();
});

test( "React isValidElement does not accept Fragments, strings, numbers, nulls, true, false, undefined", () => {

    expect([
        React.Fragment,
        10,
        0,
        "aaa",
        "",
        null,
        true,
        false,
        undefined
    ].some( React.isValidElement )).toBeFalsy();
});