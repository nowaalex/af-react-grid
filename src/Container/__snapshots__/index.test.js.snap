// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allows rendering of bools, undefined, null and empty strings 1`] = `
<div
  className="css-16ehm7c"
/>
`;

exports[`Empty container renders correctly 1`] = `
<div
  className="css-16ehm7c"
/>
`;
